Created the form for join waitlist

SebastianHirsch opened the pull request:
Closes #110

A visitor can now join the waitlist by entering his email address in the respective form.

If the visitor enters a valid email address and submits the form, a popup window informs the visitor that he successfully joined the waitlist.

![Join Waitlist Success](https://github.com/iamlogand/republic-of-rome-online/assets/4649153/96bf4e00-9371-4065-b534-f6d774d96216)

If the visitor enters an invalid email address, the helper text asks him to enter a valid one.

![Join Waitlist Fail](https://github.com/iamlogand/republic-of-rome-online/assets/4649153/e3102a9b-86f2-4acc-80e3-e9d7c5e43a12)

iamlogand commented on frontend/pages/index.tsx:
```diff
@@ -22,6 +46,19 @@ const HomePage = () => {
+      <h3 id="waitlist">Join our waitlist</h3>
+        <form onSubmit={handleSubmit}>
+          <Stack alignItems="start" spacing={2}>
+            <TextField required
+                id="email"
+                label="Email"
+                onChange={handleEmailChange}
+                helperText={"johndoe@example.com"}
+                style={{width: "300px"}} />
```
It would be good to use the `error` attribute to show a feedback message if there is any. This message will be captured by the `requestWithoutAuthentication` function already, you just need to extract it from `response` in a similar manner to frontend/pages/games/new.tsx.

SebastianHirsch replied:
This was very interesting. Thank you for pointing me in the right direction.

iamlogand commented on frontend/pages/index.tsx:
```diff
@@ -13,7 +17,27 @@ import ExternalLink from "@/components/ExternalLink";
+
+    if (response) {
+      if (response.status === 201) {
+        await router.push('/');
```
Instead of `router.push`, let's do something more useful to show the user when their form submission was a success. I suggest trying the [MUI Snackbar component](https://mui.com/material-ui/react-snackbar/), which is the MUI version of a "toast".

SebastianHirsch replied:
Very good idea and thanks for the pointer to the Snackbar component. I implemented it and think this will need some extra attention of yours when reviewing. I hope, I implemented it as you wanted it.

iamlogand commented on frontend/pages/index.tsx:
```diff
@@ -13,7 +17,27 @@ import ExternalLink from "@/components/ExternalLink";
+
+    if (response) {
+      if (response.status === 201) {
+        await router.push('/');
```
If successful, we should clear the text field by setting `email` to `""`.

SebastianHirsch replied:
Implemented that as well.

iamlogand commented on frontend/pages/index.tsx:
```diff
@@ -13,7 +17,27 @@ import ExternalLink from "@/components/ExternalLink";
+  const handleSubmit = async (event: React.ChangeEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    
+    const waitlistEntryData = { email };
```
You could remove `waitlistEntryData` and instead pass `email` directly to the request function.

SebastianHirsch replied:
A good point. I changed it

iamlogand commented on frontend/pages/index.tsx:
```diff
@@ -22,6 +46,19 @@ const HomePage = () => {
       <h2>Welcome to Republic of Rome Online</h2>
       <p><i>Experience the intrigue and power struggles of Ancient Rome, right from your browser</i></p>
 
+      <h3 id="waitlist">Join our waitlist</h3>
+        <form onSubmit={handleSubmit}>
+          <Stack alignItems="start" spacing={2}>
+            <TextField required
+                id="email"
+                label="Email"
+                onChange={handleEmailChange}
+                helperText={"johndoe@example.com"}
+                style={{width: "300px"}} />
+            <Button variant="contained" type="submit">Join waitlist</Button>
+          </Stack>
+        </form>
```
Could you move the waitlist to under the Early Development Notice and Wiki, but above the Exclusive Features bit?

SebastianHirsch replied:
I've done it and think it's a good choice of yours. It now doesn't distract too much from the Early Development Notice and the Wiki.

iamlogand reviewed (changes requested):
Looking really good! Just a few changes needed.

Please can you add a description to the PR and, if you have time, a screenshot of the new UI feature?

SebastianHirsch commented on frontend/pages/index.tsx:
```diff
@@ -22,6 +46,19 @@ const HomePage = () => {
+                id="email"
+                label="Email"
+                onChange={handleEmailChange}
+                helperText={"johndoe@example.com"}
```
Indeed not needed. I deleted it.

iamlogand commented on frontend/functions/request.ts:
```diff
@@ -109,7 +109,7 @@ export async function requestWithoutAuthentication(
   data?: object
 ): Promise<ResponseType> {
-  const baseUrl = process.env.NEXT_PUBLIC_BACKEND_ORIGIN + '/rorapp/api/';
+  const baseUrl = process.env.NEXT_PUBLIC_BACKEND_ORIGIN + '/api/';
```
This is needed:

```suggestion
  const baseUrl = process.env.NEXT_PUBLIC_API_URL;
```

Also, I suggest deleting the `NEXT_PUBLIC_BACKEND_ORIGIN` environment variable from your development environment, since it has been removed in production.

SebastianHirsch replied:
I updated it to the newest URL and it works now.

iamlogand commented on frontend/functions/request.ts:
```diff
@@ -120,6 +120,7 @@ export async function requestWithoutAuthentication(
       url: requestUrl,
       data: data
     });
+    console.log('Axios: ', response);
```
```suggestion
```

SebastianHirsch replied:
Good catch, thank you!

iamlogand reviewed (changes requested):
Thanks for implementing most of the requested changes. Just the previous unresolved comments remaining, plus I've added a couple more.

Also, could you add a video and a brief description to the PR? The video can be a 10 seconds or so short GIF showing what it looks like when submitting the waitlist form, including the appearance and disapearance of the Snackbar (similar to the video on #24). I suggest using https://gifcap.dev/ for this if you don't already have a method for capturing short GIFs of the screen.

iamlogand commented:
Also I forgot to convert the PR to draft after my first review, I meant to do that.

SebastianHirsch commented:
> Thanks for implementing most of the requested changes. Just the previous unresolved comments remaining, plus I've added a couple more.
> 
> Also, could you add a video and a brief description to the PR? The video can be a 10 seconds or so short GIF showing what it looks like when submitting the waitlist form, including the appearance and disapearance of the Snackbar (similar to the video on #24). I suggest using https://gifcap.dev/ for this if you don't already have a method for capturing short GIFs of the screen.

I added a short description with two animated GIFs, one for success, one for fail. If you want any changes, please let me know.

iamlogand reviewed (approved):
Great work!
